### PR TITLE
Explorer Display for mobile & Extension Points

### PR DESCRIPTION
**Pull request by cristian-milea**
#### What problem is this solving?

Offer a new type of menu for mobileMode, providing navigation inside the menu. This is good for deep levels of categories, keeping the user focused one category at a time.

Also, added Before and After extension points.

#### How to test it?

```
"category-menu#main": {
    "props": {
      "showAllDepartments": true,
      "showSubcategories": true,
      "menuDisposition": "left",
      "mobileMode": true,
      "mobileDisplayType": "explorer"
    },
    "blocks": [
      "before-menu",
      "after-menu"
    ]
  },
  "before-menu": {
    "children": [
      "rich-text#menu-before-child"
    ]
  },
  "rich-text#menu-before-child": {
    "props": {
      "text": " --- add enything before"
    }
  },
```

[Workspace](https://cristi--elefant.myvtex.com/)

#### Describe alternatives you've considered, if any.

This type of navigation is not possible using any other IO components.

#### Related to / Depends on
* Lodash and used tree shaking imports.

### CODE REVIEW

File: messages/pt.json [hunk 1]
```diff
@@ -13,5 +13,7 @@
   "admin/editor.category-menu.sort-type.name": "Nome",
   "admin/editor.category-menu.title": "Menu de categorias",
   "store/category-menu.all-category.title": "Ver Todos",
+  "store/category-menu.explorer.all-from": "Ver todos",
+  "store/category-menu.explorer.back": "Regresar",
```
Comment (camilavcoutinho) [inline messages/pt.json:17]:
```suggestion
  "store/category-menu.explorer.back": "Voltar",
```

File: messages/es.json [hunk 2]
```diff
@@ -13,5 +13,7 @@
   "admin/editor.category-menu.sort-type.name": "Nombre",
   "admin/editor.category-menu.title": "Menú de categorías",
   "store/category-menu.all-category.title": "Ver Todos",
+  "store/category-menu.explorer.all-from": "Ver todos",
+  "store/category-menu.explorer.back": "Regresar",
```
Comment (camilavcoutinho) [inline messages/es.json:17]:
```suggestion
  "store/category-menu.explorer.back": "Atrás",
```

**Review by camilavcoutinho — CHANGES_REQUESTED** [review]

**Review by camilavcoutinho — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -1,12 +1,30 @@
...
   "devDependencies": {
     "@types/node": "^12.7.2",
-    "eslint": "^6.2.2",
-    "eslint-config-vtex": "^11.0.0",
-    "eslint-config-vtex-react": "^5.0.1",
-    "prettier": "^1.18.2",
-    "typescript": "^3.5.3"
+    "@vtex/prettier-config": "^0.3.6",
+    "eslint": "^7.14.0",
+    "eslint-config-vtex": "^12.8.11",
```
Comment (igorbrasileiro) [inline package.json:24]: Bump to our latest version: [14.x](https://github.com/vtex/typescript/blob/main/packages/eslint-config-vtex/package.json)

File: store/interfaces.json [hunk 4]
```diff
@@ -1,5 +1,16 @@
...
+  },
+  "before-menu": {
+    "component": "Extension",
+    "composition": "children",
+    "allowed": "*"
+  },
+  "after-menu": {
+    "component": "Extension",
+    "composition": "children",
+    "allowed": "*"
```
Comment (igorbrasileiro) [inline store/interfaces.json:14]: I didn't get why you do this. Can you explain?

File: react/utils/constants.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export const DEFAULT_SUBCATEGORIES_LEVELS = 1
```
Comment (igorbrasileiro) [inline react/utils/constants.js:1]: Can you migrate to typescript files? `.ts`

File: react/tsconfig.json [hunk 6]
```diff
@@ -1,4 +1,31 @@
 {
   "extends": "@vtex/tsconfig",
-  "include": ["./**/*.ts", "./**/*.tsx", "./typings/*.d.ts"]
+  "compilerOptions": {
```
Comment (igorbrasileiro) [inline react/tsconfig.json:3]: Did you write this by yourself, or `vtex setup` do?

File: react/package.json [hunk 7]
```diff
@@ -20,7 +20,15 @@
     "@types/jest": "^24.0.18",
     "@vtex/test-tools": "^0.1.2",
     "@vtex/tsconfig": "^0.1.1",
-    "apollo-client": "^2.5.1"
+    "apollo-client": "^2.5.1",
+    "vtex.category-menu": "https://cristi--elefant.myvtex.com/_v/private/typings/linked/v1/vtex.category-menu@2.16.1+build1627629726/public/@types/vtex.category-menu",
```
Comment (igorbrasileiro) [inline react/package.json:24]: This line can't be the type of your workspace. Unlink the app from your current workspace, then run `vtex setup --typings`

File: react/index.tsx [hunk 8]
```diff
@@ -58,23 +67,32 @@ const CategoryMenu = ({
       <div className={`${styles.sidebarContainer} ${styles.mobile}`}>
         <SideBar
           visible={sideBarVisible}
-          title={intl.formatMessage({
-            id: 'store/category-menu.departments.title',
-          })}
+          // title={intl.formatMessage({
+          //   id: 'store/category-menu.departments.title',
+          // })}
```
Comment (igorbrasileiro) [inline react/index.tsx:72]: Does this was intentional?

File: react/index.tsx [hunk 9]
```diff
@@ -58,23 +67,32 @@ const CategoryMenu = ({
...
+        <div
+          className="flex pa4 pointer items-center"
+          onClick={handleSidebarToggle}
+        >
           <IconMenu size={20} />
+          <p className="flex ml4">
+            {intl.formatMessage({
+              id: 'store/category-menu.departments.title',
+            })}
+          </p>
```
Comment (igorbrasileiro) [inline react/index.tsx:87]: It seems to be a breaking change, right? @victorhmp @icazevedo, can you help here?

**Review by igorbrasileiro — CHANGES_REQUESTED** [review]
Excellent job, @cristian-milea! 
As far as I understand, this PR seems to be a breaking change to our APP, since it changes where the title appears.

Also, I left some comments. Can you take a look there and help me?

**Pull request closed without merging**